starknet_devnet/predeployed_contract_wrapper.py

```diff
@@ -1,9 +1,100 @@
 """Parent for predeployed contract wrapper classes"""
+from abc import ABC, abstractmethod
 
+from starkware.starknet.testing.starknet import Starknet
+from starkware.starknet.services.api.contract_class import ContractClass
+from starkware.starknet.testing.contract import StarknetContract
 
-class PredeployedContractWrapper:
+
+class PredeployedContractWrapper(ABC):
     """Parent class for all predeployed contract wrapper classes"""
 
+    # Cannot import it because of circular imports
+    # from .starknet_wrapper import StarknetWrapper
+
+    starknet_wrapper = NotImplemented
+
+    @property
+    def address(self) -> int:
+        try:
+            return self.ADDRESS
+        except AttributeError:
+            try:
+                return self._address
+            except AttributeError:
+                raise NotImplementedError(
+                    "Predeployed contracts should provide an address attribute or an ADDRESS class attribute"
+                )
+
+    @address.setter
+    def address(self, address: int):
+        self._address = address
+
+    @property
+    def class_hash_bytes(self) -> bytes:
+        try:
+            return self._class_hash_bytes
+        except AttributeError:
+            try:
+                return self.HASH_BYTES
+            except AttributeError:
+                raise NotImplementedError(
+                    "Predeployed contracts should provide a class_hash_bytes attribute or a HASH_BYTES class attribute"
+                )
+
+    @class_hash_bytes.setter
+    def class_hash_bytes(self, class_hash_bytes: bytes):
+        self._class_hash_bytes = class_hash_bytes
+
+    @property
+    def contract_class(self) -> ContractClass:
+        try:
+            return self._contract_class
+        except AttributeError:
+            try:
+                return self.get_contract_class()
+            except AttributeError:
+                raise NotImplementedError(
+                    "Predeployed contracts should provide a contract_class attribute or get_contract_class method"
+                )
+
+    @contract_class.setter
+    def contract_class(self, contract_class):
+        self._contract_class = contract_class
+
+    async def deploy_contract(self):
+        # pylint: disable=protected-access
+        starknet: Starknet = self.starknet_wrapper.starknet
+        starknet.state.state.cache._class_hash_writes[
+            self.address
+        ] = self.class_hash_bytes
+        # replace with await starknet.state.state.deploy_contract
+        # await starknet.state.state.deploy_contract(self.address, self.class_hash_bytes)
+        # For now, it fails for fee token since the address is the same as the
```

Comment:
Fee Token is Devnet's ETH Token, it's supposed to be on the same address

Reply:
The comment is there because it's not obvious why we're manipulating the cache and not calling `state.deploy_contract`, I spent some time understanding why and figured it'll be good to have that in the comment